common/drivers/platform/stm32l4/st_usart.cc

```diff
@@ -0,0 +1,68 @@
+#include "st_usart.h"
+
+StUsart::StUsart(USART_TypeDef* base_addr, uint32_t sys_clck,
+                 uint32_t baud_rate)
+    : base_addr(base_addr), uartdiv(sys_clck / baud_rate){};
+
+bool StUsart::receive_rx(std::span<uint8_t> data)
+{
+    static size_t index = 0;
+
+    //prevent overrun
+    if (base_addr->ISR & USART_ISR_ORE)
+    {
+        base_addr->ICR |= USART_ICR_ORECF;
+    }
+
+    if ((base_addr->ISR & USART_ISR_RXNE) && index <= data.size())
+    {
+        data[index] = base_addr->RDR;
+
+        if (data[index] == '\n' || data[index] == '\r' || index > data.size())
+        {
+            index = 0;
+            return true;
+        }
+        index++;
+    }
+
+    return false;
+}
+
+bool StUsart::send_tx(const std::span<uint8_t> data)
```

Comment:
Change param type to std::span const uint8_t since we are concerned about not modifying the data inside the function.
```suggestion
bool StUsart::send_tx(std::span<const uint8_t> data)
```